packages/tokens/non-fungible/src/extensions/enumerable/storage.rs

```diff
@@ -0,0 +1,536 @@
+///
+/// # Arguments
+///
+/// * `e` - The environment reference.
+/// * `from` - The current owner's address.
+/// * `to` - The recipient's address.
+/// * `token_id` - The identifier of the token being transferred.
+///
+/// # Errors
+///
+/// * refer to [`crate::storage::transfer`] errors.
+/// * refer to [`remove_from_owner_enumeration`] errors.
+///
+/// # Events
+///
+/// * topics - `["transfer", from: Address, to: Address]`
+/// * data - `[token_id: TokenId]`
+///
+/// # Notes
+///
+/// This is a wrapper around [`crate::storage::transfer`], that also
+/// handles the storage updates for:
+/// * owner_tokens enumeration
+pub fn transfer(e: &Env, from: &Address, to: &Address, token_id: TokenId) {
+    base_transfer(e, from, to, token_id);
+
+    remove_from_owner_enumeration(e, from, token_id);
+    add_to_owner_enumeration(e, to, token_id);
+}
+
+/// Transfers a non-fungible token (NFT), ensuring ownership and approval
+/// checks.
+///
+/// # Arguments
+///
+/// * `e` - The environment reference.
+/// * `spender` - The address attempting to transfer the token.
+/// * `from` - The current owner's address.
+/// * `to` - The recipient's address.
+/// * `token_id` - The identifier of the token being transferred.
+///
+/// # Errors
+///
+/// * refer to [`crate::storage::transfer_from`] errors.
+/// * refer to [`remove_from_owner_enumeration`] errors.
+///
+/// # Events
+///
+/// * topics - `["transfer", from: Address, to: Address]`
+/// * data - `[token_id: TokenId]`
+///
+/// # Notes
+///
+/// This is a wrapper around [`crate::storage::transfer_from`], that also
+/// handles the storage updates for:
+/// * owner_tokens enumeration
+pub fn transfer_from(e: &Env, spender: &Address, from: &Address, to: &Address, token_id: TokenId) {
+    base_transfer_from(e, spender, from, to, token_id);
+
+    remove_from_owner_enumeration(e, from, token_id);
+    add_to_owner_enumeration(e, to, token_id);
+}
+// ################## LOW-LEVEL HELPERS ##################
+
+/// Returns the old total supply (before the increment).
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+///
+/// # Errors
+///
+/// * [`NonFungibleTokenError::TokenIDsAreDepleted`] - When attempting to mint a
+///   new token ID, but all token IDs are already in use.
+pub fn increment_total_supply(e: &Env) -> TokenId {
+    let total_supply = total_supply(e);
+    let Some(new_total_supply) = total_supply.checked_add(1) else {
+        panic_with_error!(e, NonFungibleTokenError::TokenIDsAreDepleted);
+    };
+    e.storage().instance().set(&StorageKey::TotalSupply, &new_total_supply);
+
+    total_supply
+}
+
+/// Returns the new total supply (after the decrement).
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+pub fn decrement_total_supply(e: &Env) -> TokenId {
+    let total_supply = total_supply(e);
+    let new_total_supply = total_supply.checked_sub(1).expect("Total supply cannot be negative");
```

Comment:
I'd suggest using `saturated_sub` ensuring it never goes negative and mentioning it in the docstring.

Reply:
That's my understanding too, we want it always >= 0 similar to `balance`, right? 🤔